plugins/3rdParty/qChaiScripting/ChaiScriptingPlugin.cpp

```diff
@@ -0,0 +1,409 @@
+	{
+		ccLog::Error(ee.what());
+		if (ee.call_stack.size() > 0)
+		{
+			ccLog::Warning(QString("during evaluation at (%1, %2)").arg(ee.call_stack[0]->start().line).arg(ee.call_stack[0]->start().column));
+		}
+		std::cout << std::endl;
+		return false;
+	}
+	catch (const std::exception & e)
+	{
+		ccLog::Error(e.what());
+		return false;
+	}
+	return true;
+}
+
+QList<QAction *> ChaiScriptingPlugin::getActions()
+{
+	if ( !m_action )
+	{
+		m_action = new QAction( getName(), this );
+		m_action->setToolTip( getDescription() );
+		m_action->setIcon( getIcon() );
+		
+		// Connect appropriate signal
+		connect( m_action, &QAction::triggered, this, &ChaiScriptingPlugin::openScriptEditor);
+	}
+
+	cseMW = chaiScriptCodeEditorMainWindow::TheInstance();
+	connect(cseMW, &chaiScriptCodeEditorMainWindow::executionCalled, this, &ChaiScriptingPlugin::executionCalled);
+	connect(cseMW, &chaiScriptCodeEditorMainWindow::reset_Chai_to_initial_state, this, &ChaiScriptingPlugin::resetToDefaultChaiState);
+	connect(cseMW, &chaiScriptCodeEditorMainWindow::save_Chai_state, this, &ChaiScriptingPlugin::saveChaiState);
+	connect(cseMW, &chaiScriptCodeEditorMainWindow::reset_chai_to_last_save, this, &ChaiScriptingPlugin::resetToSavedChaiState);
+	connect(cseMW, &chaiScriptCodeEditorMainWindow::destroy_chai, this, &ChaiScriptingPlugin::destroyChai);
+
+	return { m_action };
+}
+
+void ChaiScriptingPlugin::stop()
+{
+	destroyChai();
+}
+
+void ChaiScriptingPlugin::openScriptEditor()
+{
+	
+	if (!cseMW)
+	{
+		cseMW = chaiScriptCodeEditorMainWindow::TheInstance();
+	}
+	if (!cseMW)
+	{
+		m_app->dispToConsole("Failed to open chaiScriptCodeEditorMainWindow", ccMainAppInterface::ERR_CONSOLE_MESSAGE);
+		return;
+	}
+	cseMW->show();
+	cseMW->raise();
+	cseMW->activateWindow();
+	if (!chai)
+	{
+		setupChaiScriptEngine();
+	}
+
+}
+
+void ChaiScriptingPlugin::registerCommands(ccCommandLineInterface* cmd)
+{	
+	if (!cmd)
+	{
+		assert(false);
+		return;
+	}
+	if (!chai)
+	{
+		setupChaiScriptEngine();
+	}
+	storedCMDLine = cmd;
+	chai->add_global(var(cmd), "m_cmd");
+	chai->eval(R"(global print = fun(x){m_cmd.print(x.to_QString());})");
+	cmd->registerCommand(ccCommandLineInterface::Command::Shared(new CommandCHAI()));	
+}
+
+void ChaiScriptingPlugin::registerCommand(CommandCHAIDerived newCmd)
+{
+	if (storedCMDLine)
+	{
+		ccLog::Print(QString("Attempting to register %1 : %2").arg(newCmd.m_keyword).arg(newCmd.m_name));
+		if (storedCMDLine->registerCommand(ccCommandLineInterface::Command::Shared(new CommandCHAIDerived(newCmd.m_name, newCmd.m_keyword.toUpper(), newCmd.chai_process))))
+		{
+			ccLog::Print("registerCommand success!");
+		}
+		else
+		{
+			ccLog::Error("registerCommand Failed");
+		}
+	}
+}
+
+void ChaiScriptingPlugin::onNewSelection(const ccHObject::Container& selectedEntities)
+{
+	if (chai_onNewSelection)
+	{
+		std::vector<std::shared_ptr<ccHObject>> sel;
+		try
+		{
+			for each (auto objectName in selectedEntities)
```

Comment:
This doesn't look like standard C++ 😧

Reply:
yeah I'm living in MSVC land :( I'll get this patched up